### PR TITLE
Odyssey: store the timestamp when enabling / disabling new Stats

### LINKED ISSUE #28709: Odyssey Stats: Make Odyssey Stats enabled by default
This change is part of our March release strategy; refer to p1HpG7-khi-p2 for more details.

All existing customers should be opted into Odyssey Stats unless they've explicitly opted out, which should be detectable by checking for a `false` option value.

### PR DESCRIPTION
**Pull request by kangzj**
## Proposed changes:
The PR proposes to 

- Add new stats options to store the timestamp when Odyssey is enabled / disabled, so that we could implement the "This prompt should be shown a month after the customer has opted out of Odyssey Stats" in #28729.
- Add `collapse_nudges` to whitelist, otherwise it'd be cleared on saving.
- Renamed stats option `enable_calypso_stats` to `enable_odyssey_stats`
- *Default `enable_odyssey_stats` to true, meaning for Jetpack sites old or new, Odyssey would be the default experience‼️* 

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
n/a

## Does this pull request change what data or activity we track or use?
no

## Testing instructions:

* Go to `/wp-admin/admin.php?page=jetpack#/traffic` and find section `Jetpack Stats`
* Toggle `Enable the new Jetpack Stats experience New`
* Open a MySQL querying tool as you like, for example PHPMyAdmin
* Run SQL: `SELECT * FROM `wp_options` WHERE `option_name` LIKE 'stats_options'`
* Unserialize `option_value` with [online tool](https://www.unserialize.com/) or locally with PHP CLI
* Ensure 'collapse_nudges', 'enable_calypso_stats', 'calypso_stats_last_enabled_at', 'calypso_stats_last_disabled_at' changes as expected.

### CODE REVIEW

File: projects/packages/stats/tests/php/test-xmlrpc-provider.php [hunk 1]
```diff
@@ -92,23 +92,27 @@ public function test_get_blog() {
...
-			'stats_api'           => 'jetpack',
-			'page_on_front'       => false,
-			'permalink_structure' => false,
-			'category_base'       => false,
-			'tag_base'            => false,
+			'admin_bar'                      => true,
+			'count_roles'                    => array(),
+			'do_not_track'                   => true,
+			'version'                        => Main::STATS_VERSION,
+			'collapse_nudges'                => false,
```
Comment (kangzj) [inline projects/packages/stats/tests/php/test-xmlrpc-provider.php]: added these 4 lines, and others are all formatted

File: projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 2]
```diff
@@ -863,21 +864,31 @@ public function update_data( $request ) {
 					break;
 
 				case 'admin_bar':
-				case 'enable_calypso_stats':
 				case 'roles':
 				case 'count_roles':
 				case 'blog_id':
-				case 'do_not_track':
 				case 'version':
+					$updated = Stats_Options::set_option( $option, $value );
```
Comment (kangzj) [inline projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Refactored to use `Stats_Options`

File: projects/packages/stats/src/class-options.php [hunk 3]
```diff
@@ -155,12 +155,16 @@ public static function upgrade_options( $options ) {
...
-			'do_not_track' => true, // @todo
-			'blog_id'      => Jetpack_Options::get_option( 'id' ),
-			'version'      => Main::STATS_VERSION,
+			'admin_bar'                      => true,
+			'roles'                          => array( 'administrator' ),
+			'count_roles'                    => array(),
+			'do_not_track'                   => true, // @todo
+			'blog_id'                        => Jetpack_Options::get_option( 'id' ),
+			'version'                        => Main::STATS_VERSION,
+			'collapse_nudges'                => false,
```
Comment (kangzj) [inline projects/packages/stats/src/class-options.php]: Adds default values which are also used as a whitelist.

File: projects/packages/stats/tests/php/test-options.php [hunk 4]
```diff
@@ -33,14 +33,18 @@ public function tear_down() {
 	public function test_get_options() {
```
Comment (kangzj) [inline projects/packages/stats/tests/php/test-options.php:33]: Adds defaults for the tests to pass

File: projects/packages/stats/src/class-options.php [hunk 5]
```diff
@@ -155,12 +155,15 @@ public static function upgrade_options( $options ) {
...
-			'blog_id'      => Jetpack_Options::get_option( 'id' ),
-			'version'      => Main::STATS_VERSION,
+			'admin_bar'                => true,
+			'roles'                    => array( 'administrator' ),
+			'count_roles'              => array(),
+			'do_not_track'             => true, // @todo
+			'blog_id'                  => Jetpack_Options::get_option( 'id' ),
+			'version'                  => Main::STATS_VERSION,
+			'collapse_nudges'          => false,
+			'enable_calypso_stats'     => false,
```
Comment (jsnmoon) [inline projects/packages/stats/src/class-options.php]:
```suggestion
			'enable_calypso_stats'     => true,
```
Could we set the default to true here and address https://github.com/Automattic/jetpack/issues/28709?

Also, what are your thoughts on referencing Odyssey instead of Calypso in the option name (e.g., `enable_odyssey_stats`)? I understand this might not be realistic due to backward compatibility concerns.

File: projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 6]
```diff
@@ -863,21 +866,36 @@ public function update_data( $request ) {
 					break;
 
 				case 'admin_bar':
-				case 'enable_calypso_stats':
 				case 'roles':
 				case 'count_roles':
 				case 'blog_id':
-				case 'do_not_track':
```
Comment (jsnmoon) [inline projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: What's the benefit of this change? It appears out of scope and has a potential for regression.

**Review by jsnmoon — DISMISSED** [review]
Changes work as expected. I tested this PR using the following command:

```bash
jetpack docker wp db query "SELECT * FROM wp_options WHERE option_name like 'stats_options'"
```

File: projects/packages/stats/src/class-options.php (hunk 5 above)
Comment (kangzj) [reply projects/packages/stats/src/class-options.php]: @jsnmoon Ya two good points. I was planning to address #28709 with `enable_odyssey_stats` default to `true` for the all-in roll out.  

> I understand this might not be realistic due to backward compatibility concerns.

I'm not too concerned about this one since we haven't launched yet 😄 What do you think?

File: projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 6 above)
Comment (kangzj) [reply projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: True. I'll revert the change here, and only move `enable_calypso_stats` to limit the scope.

File: projects/packages/stats/src/class-options.php (hunk 5 above)
Comment (jsnmoon) [reply projects/packages/stats/src/class-options.php]: Let's rename it to `enable_odyssey_stats` then 🙂

File: projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 2 above)
Comment (kangzj) [reply projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: reverted

File: projects/packages/stats/src/class-options.php (hunk 5 above)
Comment (kangzj) [reply projects/packages/stats/src/class-options.php]: Renamed in 60b7e07e54

File: projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 6 above)
Comment (kangzj) [reply projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Changes reverted in fd309ea265

File: projects/packages/stats/src/class-options.php [hunk 7]
```diff
@@ -155,12 +155,15 @@ public static function upgrade_options( $options ) {
...
-			'version'      => Main::STATS_VERSION,
+			'admin_bar'                => true,
+			'roles'                    => array( 'administrator' ),
+			'count_roles'              => array(),
+			'do_not_track'             => true, // @todo
+			'blog_id'                  => Jetpack_Options::get_option( 'id' ),
+			'version'                  => Main::STATS_VERSION,
+			'collapse_nudges'          => false,
+			'enable_odyssey_stats'     => true,
+			'calypso_stats_changed_at' => 0,
```
Comment (jsnmoon) [inline projects/packages/stats/src/class-options.php]:
```suggestion
			'odyssey_stats_changed_at' => 0,
```

**Review by jsnmoon — COMMENTED** [review]
The overall behavior looks good, but you missed renaming `calypso_stats_changed_at` to `odyssey_stats_changed_at` 🙂

File: projects/packages/stats/src/class-options.php (hunk 7 above)
Comment (jsnmoon) [reply projects/packages/stats/src/class-options.php]: Fixed in bf45b96.

**Review by jsnmoon — APPROVED** [review]
I pushed up a renamed option (`odyssey_stats_changed_at`) and a trunk merge. Everything appears to work as expected.

**Comment by kangzj** [comment]
@jsnmoon Thanks for helping shepherd this one 🙌
